src/Data/SnocList.idr

```diff
@@ -0,0 +1,76 @@
+||| A Reversed List
+module Data.SnocList
+
+import Decidable.Equality
+%default total
+
+public export
+data SnocList : (type : Type) -> Type where
+  Empty : SnocList type
+  Snoc  : SnocList type -> type -> SnocList type
+
+%name SnocList sx, sy, sz
+
+||| Reverse the list and make a list.
+public export
+toList : SnocList type -> List type
+toList Empty = Nil
+toList (Snoc sx x) = x :: toList sx
+
+||| Transform to a list but keeping the contents in the 'correct' order.
+public export
+asList : SnocList type -> List type
+asList Empty = Nil
+asList (Snoc sx x) = asList sx ++ [x]
+
+
+public export
+Eq a => Eq (SnocList a) where
+  (==) Empty Empty = True
+  (==) (Snoc sx x) (Snoc sy y) = sx == sy && x == y
```

Comment:
`(&&)` is lazy in its second argument so it's probably better to do the comparisons in the other order to save exploring the tails if the heads already don't match

Reply:
made changes and forced pushed